File: aoc_day1/src/bin/part2.py
# open ./input2.txt and store it in a list
with open("input2.txt") as f:
    content = f.readlines()
content = [x.strip() for x in content]

num_map = {
    "one": 1,
    "two": 2,
    "three": 3,
    "four": 4,
    "five": 5,
    "six": 6,
    "seven": 7,
    "eight": 8,
    "nine": 9,
}
res = []
for line in content:
    tmp = []
    for i in range(len(line)):
        if line[i].isdigit():
            tmp.append(int(line[i]))
        for j in range(i + 1, len(line)):
            if line[i : j + 1] in num_map:
                tmp.append(num_map[line[i : j + 1]])

    res.append(tmp)
ans = 0
for line in res:
    num = line[0] * 10 + line[-1]
    ans += num
print(ans)

File: aoc_day1/src/bin/part2.rs
fn main() {
    let input = include_str!("./input2.txt");
    let output = part2(input);
    dbg!(output);
}

fn part2(input: &str) -> i32 {
    let mut res = 0;
    for line in input.lines() {
        let line = line.replace("one", "o1e")
            .replace("two", "t2o")
            .replace("three", "t3e")
            .replace("four", "f4r")
            .replace("five", "f5e")
            .replace("six", "s6x")
            .replace("seven", "s7n")
            .replace("eight", "e8t")
            .replace("nine", "n9e");
        // convert lines to list of chars
        let chars: Vec<char> = line.chars().collect();
        // remove all non-numeric chars
        let nums: Vec<char> = chars.into_iter().filter(|c| c.is_numeric()).collect();
        // get first and last items and convert to i32
        let first = nums.first().unwrap().to_digit(10).unwrap() as i32;
        let last = nums.last().unwrap().to_digit(10).unwrap() as i32;
        let num = first * 10 + last;

        res += num;
    }
    res
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test2() {
        let result = part2("two1nine
        eightwothree
        abcone2threexyz
        xtwone3four
        4nineeightseven2
        zoneight234
        7pqrstsixteen");
        assert_eq!(result, 281);
    }
}